lib/tests/streamlit/hashing_test.py

```diff
@@ -352,6 +352,56 @@ def test_override_streamlit_hash_func(self):
         hash_funcs = {int: lambda x: "hello"}
         self.assertNotEqual(get_hash(1), get_hash(1, hash_funcs=hash_funcs))
 
+    def test_sqlalchemy_engine(self):
+        import sqlalchemy as db
+
+        def connect():
+            pass
+
+        def connect_with_args(user):
+            pass
+
+        url = "postgresql://localhost/db"
+        auth_url = "postgresql://user:pass@localhost/db"
+
+        # Engine.url
+        self.assertEqual(
+            get_hash(db.create_engine(url)), get_hash(db.create_engine(url))
+        )
+
+        # Engine.url
+        self.assertNotEqual(
+            get_hash(db.create_engine(url)), get_hash(db.create_engine(auth_url))
+        )
+
+        # Engine.dialect
+        self.assertNotEqual(
+            get_hash(db.create_engine(url, encoding="utf-8")),
+            get_hash(db.create_engine(url, encoding="ascii")),
+        )
+
+        # Engine.pool._creator
+        self.assertNotEqual(
+            get_hash(db.create_engine(auth_url)),
+            get_hash(db.create_engine(auth_url, creator=connect)),
+        )
+
+        # BAD: We don't have access to the connect_args so we will
```

Comment:
I think it's fine to leave this commented-out block as documentation; my only question is the "BAD" tag. What does this mean to future readers of the code? Is it a TODO, a heads-up, or...?

Reply:
Well, perhaps I could explain that a bit better and change it to `NOTE`

We could have two engines that connect with different usernames, or even to different databases entirely, but if they do so using `connect_args`, their hash could be equal because we aren't hashing `connect_args`.

Reply:
Why not add these tests anyway? So if the behavior ever changes at least we'll know?

(And leave comments in the tests saying this is just how things work, but not how we _wished_ they worked)